includes/Hooks.php

```diff
@@ -148,6 +152,122 @@ public static function factory(
+	 * @param string[] &$attribs Link attributes
+	 * @param string &$ret The value to return if the hook returns false
+	 * 
+	 * @return bool
+	 */
+	public function onHtmlPageLinkRendererEnd(
+		$linkRenderer,
+		$target,
+		$isKnown,
+		&$text,
+		&$attribs,
+		&$ret
+	) {
+		if ( $isKnown || $target->isExternal() ) {
+			return true;
+		}
+
+		$attribs['rel'] = 'nofollow';
+
+		return true;
+	}
+
+	/**
+	 * Add noindex to some pages for SEO purposes. Indexing these pages is bad for SEO because
+	 * it wastes crawl budget, so we'll reduce the indexing of such pages.
+	 * 
+	 * @see https://ahrefs.com/blog/content-pruning/
+	 * @see https://ahrefs.com/blog/crawl-budget/
+	 * @see https://gitlab.com/hydrawiki/extensions/seo/-/blob/master/SEOHooks.php?ref_type=heads
+	 * 
+	 * @param OutputPage $out The OutputPage object
+	 * @param Skin $skin The Skin object that will be used to generate the page
+	 */
+	public function onBeforePageDisplay( $out, $skin ): void {
+
+		// If something has absolutely no value to someone searching on Google, then inclue it in $noIndexNamespaces
+		// For example, a wiki's CSS styles, a user's user page, the talk page of an article, etc. do not need to appear in Google,
+		// because they are not something you'd actually look up to find any kind of useful information.
+		// 
+		// I am tempted to add templates and modules here, but I believe some wikis, like devwiki, may want their templates to be indexed.
+		// I also am not including every possible talk page here, because some wikis might have troubleshooting instructions or other interesting things there.
+		// 
+		// In general, if you can think of any wiki where it might be useful to index a namespace, it's best not to add it here.
+		// 
+		// TODO: Change ManageWiki so that some namespaces default to noindex, then remove this.
+
+		$noIndexNamespaces = [
+			NS_SPECIAL,
+			NS_CATEGORY_TALK,
+			NS_MEDIAWIKI,
+			NS_MEDIAWIKI_TALK,
+			NS_USER,
+			NS_USER_TALK,
+			NS_TALK,
+			NS_FILE_TALK,
+			NS_PROJECT_TALK,
+			NS_TEMPLATE_TALK
+		];
+
+		if ( in_array( $out->getTitle()->getNamespace(), $noIndexNamespaces ) ) {
+			$out->setRobotPolicy( 'noindex,nofollow' );
+			return;
+		}
+
+		$noIndexURLParamKeys = [
+			'action',
+			'curid',
+			'diff',
+			'from',
+			'group',
+			'mobileaction',
+			'oldid',
+			'printable',
+			'profile',
+			'redirect',
+			'redlink',
+			'stableid',
+			'veaction',
+		];
+
+		$noIndexURLParamKeyValuePairs = [
+			'feed' => [ 'rss' ],
+			'limit' => [ '500' ],
+			'title' => [
+				'Category:Noindexed_pages',
+				'Category:Hidden_categories',
+			],
+		];
+
+		foreach ( $out->getRequest()->getValues() as $key => $value ) {
+			if ( in_array( $key, $noIndexURLParamKeys ) ) {
+				$out->setRobotPolicy( 'noindex,nofollow' );
+				return;
+			}
+
+			if ( in_array( $value, $noIndexURLParamKeyValuePairs[$key] ?? [] ) ) {
+				$out->setRobotPolicy( 'noindex,nofollow' );
+				return;
+			}
+		}
+	}
```

Comment:
`noIndexNamespaces` is going to be moved to a separate extension, so that review no longer applies. However, for your advice about using `array_intersect`, could you please elaborate on the reasoning?